File: packages/android/src/main/java/com/kozmos/components/MapControlButton/MapControlButton.kt
package com.kozmos.components.mapcontrolbutton

import androidx.compose.animation.animateColorAsState
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosThemeTokens
import com.kozmos.tokens.KozmosDimensions

enum class KozmosMapControlButtonPresentation {
    IconOnly,
    Labelled
}

/**
 * How an active map control reads.
 *
 * [Tinted] keeps the map surface and colours the icon and the edge, which is
 * what the SDK draws — a control over a map has to stay legible against the
 * tiles behind it, and a solid fill hides the very thing it sits on.
 * [Filled] is the inverted treatment this component shipped before
 * 2026-09-15, kept for callers that want the heavier emphasis.
 */
enum class KozmosMapControlButtonEmphasis {
    Tinted,
    Filled
}

/**
 * Where a control's state sits relative to its label.
 *
 * [Inline] runs them along one line. [Stacked] sets the state under the label,
 * which is how a map pill fits a two-word state into a control that has to
 * stay thumb-sized.
 */
enum class KozmosMapControlButtonLabelPlacement {
    Inline,
    Stacked
}

/**
 * What a map control's state resolves to, before any colour is chosen.
 *
 * Kept apart from the composable because this decision is the part a design
 * ruling changes — tinted became the default on 2026-09-15 — and it can be
 * tested without rendering anything.
 */
internal data class KozmosMapControlButtonAppearance(
    val surface: Surface,
    val icon: Tone,
    val label: Tone,
    /** The small line above a stacked state; muted only on the map's surface. */
    val caption: Tone,
    val edge: Edge
) {
    enum class Surface { Chrome, Filled }
    enum class Tone { Ink, Muted, Theme, OnFill }
    enum class Edge { Subtle, Theme }

    companion object {
        fun resolve(
            pressed: Boolean,
            emphasis: KozmosMapControlButtonEmphasis
        ): KozmosMapControlButtonAppearance = when {
            pressed && emphasis == KozmosMapControlButtonEmphasis.Filled ->
                KozmosMapControlButtonAppearance(Surface.Filled, Tone.OnFill, Tone.OnFill, Tone.OnFill, Edge.Subtle)
            // Only the glyph and the edge take the theme, so the label keeps its
            // contrast against a surface that stays the map's.
            pressed ->
                KozmosMapControlButtonAppearance(Surface.Chrome, Tone.Theme, Tone.Ink, Tone.Muted, Edge.Theme)
            else ->
                KozmosMapControlButtonAppearance(Surface.Chrome, Tone.Ink, Tone.Ink, Tone.Muted, Edge.Subtle)
        }
    }
}

@Composable
@ReadOnlyComposable
private fun KozmosMapControlButtonAppearance.Tone.color() = when (this) {
    KozmosMapControlButtonAppearance.Tone.Ink -> KozmosThemeTokens.primitivesColorsForeground100
    KozmosMapControlButtonAppearance.Tone.Muted -> KozmosThemeTokens.primitivesColorsForeground400
    KozmosMapControlButtonAppearance.Tone.Theme -> KozmosThemeTokens.primitivesColorsTheme600
    KozmosMapControlButtonAppearance.Tone.OnFill -> KozmosThemeTokens.componentsPrimaryButtonsThemedButtonForegroundContentIdle
}

/**
 * A single floating map control.
 *
 * Mirrors the React `MapControlButton`. [label] is the localized action name
 * and always becomes the accessible name; [stateLabel] is appended so screen
 * reader users hear the current state without relying on visual styling.
 */
@Composable
fun KozmosMapControlButton(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    icon: (@Composable () -> Unit)? = null,
    stateLabel: String? = null,
    presentation: KozmosMapControlButtonPresentation = KozmosMapControlButtonPresentation.IconOnly,
    emphasis: KozmosMapControlButtonEmphasis = KozmosMapControlButtonEmphasis.Tinted,
    labelPlacement: KozmosMapControlButtonLabelPlacement = KozmosMapControlButtonLabelPlacement.Inline,
    pressed: Boolean = false,
    enabled: Boolean = true
) {
    val accessibleLabel = if (stateLabel != null) "$label, $stateLabel" else label
    val appearance = KozmosMapControlButtonAppearance.resolve(pressed, emphasis)
    val borderColor by animateColorAsState(
        targetValue = if (appearance.edge == KozmosMapControlButtonAppearance.Edge.Theme) {
            KozmosThemeTokens.primitivesColorsTheme600
        } else {
            KozmosThemeTokens.primitivesColorsForeground300
        },
        label = "MapControlButtonBorder"
    )

    Surface(
        onClick = onClick,
        modifier = modifier
            .height(44.dp)
            .then(
                if (presentation == KozmosMapControlButtonPresentation.IconOnly) {
                    Modifier.size(44.dp)
                } else {
                    Modifier.widthIn(max = 256.dp).defaultMinSize(minWidth = 44.dp)
                }
            )
            .semantics {
                contentDescription = accessibleLabel
                selected = pressed
            },
        enabled = enabled,
        shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusControl),
        color = if (appearance.surface == KozmosMapControlButtonAppearance.Surface.Filled) {
            KozmosThemeTokens.componentsPrimaryButtonsThemedButtonBackgroundIdle
        } else {
            KozmosThemeTokens.primitivesColorsBackground0.copy(alpha = 0.9f)
        },
        contentColor = appearance.label.color(),
        border = BorderStroke(1.dp, borderColor),
        shadowElevation = if (pressed) 2.dp else 8.dp
    ) {
        Row(
            modifier = Modifier.padding(
                horizontal = if (presentation == KozmosMapControlButtonPresentation.Labelled) {
                    KozmosDimensions.primitivesLayoutSpacing150
                } else {
                    0.dp
                }
            ),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(
                KozmosDimensions.primitivesLayoutSpacing100,
                Alignment.CenterHorizontally
            )
        ) {
            if (icon != null) {
                CompositionLocalProvider(LocalContentColor provides appearance.icon.color()) {
                    icon()
                }
            }

            if (presentation == KozmosMapControlButtonPresentation.Labelled) {
                if (labelPlacement == KozmosMapControlButtonLabelPlacement.Stacked) {
                    Column(horizontalAlignment = Alignment.Start) {
                        // The SDK sets these at 11sp over 13sp semibold. The
                        // type scale has no role at either size yet, so this
                        // reaches for the nearest roles and the deviation is
                        // recorded in the gap list rather than hard-coded here.
                        Text(
                            text = label,
                            style = MaterialTheme.typography.labelSmall,
                            color = appearance.caption.color(),
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )

                        if (stateLabel != null) {
                            Text(
                                text = stateLabel,
                                style = MaterialTheme.typography.bodySmall,
                                fontWeight = FontWeight.SemiBold,
                                maxLines = 1
                            )
                        }
                    }
                } else {
                    Text(
                        text = label,
                        style = MaterialTheme.typography.bodyMedium,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis
                    )

                    if (stateLabel != null) {
                        Text(
                            text = stateLabel,
                            style = MaterialTheme.typography.bodyMedium,
                            fontWeight = FontWeight.SemiBold,
                            maxLines = 1
                        )
                    }
                }
            }
        }
    }
}

File: packages/ios/Sources/Components/MapControlButton/MapControlButton.swift
import SwiftUI

public enum KozmosMapControlButtonPresentation {
    case iconOnly
    case labelled
}

/// How an active map control reads.
///
/// `tinted` keeps the map surface and colours the icon and the edge, which is
/// what the SDK draws — a control over a map has to stay legible against the
/// tiles behind it, and a solid fill hides the very thing it sits on.
/// `filled` is the inverted treatment this component shipped before
/// 2026-09-15, kept for callers that want the heavier emphasis.
public enum KozmosMapControlButtonEmphasis {
    case tinted
    case filled
}

/// Where a control's state sits relative to its label.
///
/// `inline` runs them along one line. `stacked` sets the state under the
/// label, which is how a map pill fits a two-word state into a control that
/// has to stay thumb-sized.
public enum KozmosMapControlButtonLabelPlacement {
    case inline
    case stacked
}

/// What a map control's state resolves to, before any colour is chosen.
///
/// Kept apart from the view because this decision is the part a design ruling
/// changes — tinted became the default on 2026-09-15 — and it can be tested
/// without rendering anything.
struct KozmosMapControlButtonAppearance: Equatable {
    enum Surface: Equatable { case chrome, filled }
    enum Tone: Equatable { case ink, muted, theme, onFill }
    enum Edge: Equatable { case subtle, theme }

    let surface: Surface
    let icon: Tone
    let label: Tone
    /// The small line above a stacked state. Muted on the map's surface, but on
    /// a filled one a muted grey would sit at about 1.9:1 against the theme.
    let caption: Tone
    let edge: Edge

    init(pressed: Bool, emphasis: KozmosMapControlButtonEmphasis) {
        switch (pressed, emphasis) {
        case (true, .filled):
            surface = .filled
            icon = .onFill
            label = .onFill
            caption = .onFill
            edge = .subtle
        case (true, .tinted):
            // Only the glyph and the edge take the theme, so the label keeps
            // its contrast against a surface that stays the map's.
            surface = .chrome
            icon = .theme
            label = .ink
            caption = .muted
            edge = .theme
        case (false, _):
            surface = .chrome
            icon = .ink
            label = .ink
            caption = .muted
            edge = .subtle
        }
    }
}

/// A single floating map control.
///
/// Mirrors the React `MapControlButton`. `label` is the localized action name
/// and always becomes the accessible name; `stateLabel` is appended so screen
/// reader users hear the current state without relying on visual styling.
public struct KozmosMapControlButton<Icon: View>: View {
    private let icon: Icon
    private let label: String
    private let stateLabel: String?
    private let presentation: KozmosMapControlButtonPresentation
    private let emphasis: KozmosMapControlButtonEmphasis
    private let labelPlacement: KozmosMapControlButtonLabelPlacement
    private let pressed: Bool
    private let isDisabled: Bool
    private let action: () -> Void

    /// Someone who has asked iOS to reduce motion still needs to read the new
    /// state; they just should not watch the control grow to show it.
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    public init(
        label: String,
        stateLabel: String? = nil,
        presentation: KozmosMapControlButtonPresentation = .iconOnly,
        emphasis: KozmosMapControlButtonEmphasis = .tinted,
        labelPlacement: KozmosMapControlButtonLabelPlacement = .inline,
        pressed: Bool = false,
        isDisabled: Bool = false,
        action: @escaping () -> Void,
        @ViewBuilder icon: () -> Icon
    ) {
        self.label = label
        self.stateLabel = stateLabel
        self.presentation = presentation
        self.emphasis = emphasis
        self.labelPlacement = labelPlacement
        self.pressed = pressed
        self.isDisabled = isDisabled
        self.action = action
        self.icon = icon()
    }

    private var accessibleLabel: String {
        guard let stateLabel else { return label }
        return "\(label), \(stateLabel)"
    }

    var appearance: KozmosMapControlButtonAppearance {
        KozmosMapControlButtonAppearance(pressed: pressed, emphasis: emphasis)
    }

    private func color(_ tone: KozmosMapControlButtonAppearance.Tone) -> Color {
        switch tone {
        case .ink:
            return KozmosColors.primitivesColorsForeground100
        case .muted:
            return KozmosColors.primitivesColorsForeground400
        case .theme:
            return KozmosColors.primitivesColorsTheme600
        case .onFill:
            return KozmosColors.componentsPrimaryButtonsThemedButtonForegroundContentIdle
        }
    }

    private var backgroundColor: Color {
        appearance.surface == .filled
            ? KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle
            : KozmosColors.primitivesColorsBackground0.opacity(0.9)
    }

    private var borderColor: Color {
        appearance.edge == .theme
            ? KozmosColors.primitivesColorsTheme600
            : KozmosColors.primitivesColorsForeground300
    }

    @ViewBuilder
    private var labelContent: some View {
        if presentation == .labelled {
            if labelPlacement == .stacked {
                VStack(alignment: .leading, spacing: 0) {
                    // The SDK sets these at 11pt over 13pt semibold. The type
                    // scale has no role at either size yet, so this reaches for
                    // the nearest roles and the deviation is recorded in the
                    // gap list rather than hard-coded here.
                    Text(label)
                        .font(KozmosTypography.caption)
                        .foregroundColor(color(appearance.caption))
                        .lineLimit(1)
                        .truncationMode(.tail)

                    if let stateLabel {
                        Text(stateLabel)
                            .font(KozmosTypography.footnote)
                            .fontWeight(.semibold)
                            .lineLimit(1)
                    }
                }
            } else {
                Text(label)
                    .font(KozmosTypography.subheadline)
                    .fontWeight(.medium)
                    .lineLimit(1)
                    .truncationMode(.tail)

                if let stateLabel {
                    Text(stateLabel)
                        .font(KozmosTypography.subheadline)
                        .fontWeight(.semibold)
                        .lineLimit(1)
                }
            }
        }
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                icon
                    .foregroundColor(color(appearance.icon))
                    .accessibilityHidden(true)

                labelContent
            }
            .foregroundColor(color(appearance.label))
            .frame(
                width: presentation == .iconOnly ? 44 : nil,
                height: 44
            )
            .frame(maxWidth: presentation == .labelled ? 256 : nil)
            .padding(.horizontal, presentation == .labelled ? KozmosDimensions.primitivesLayoutSpacing150 : 0)
            .background(backgroundColor)
            .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl, style: .continuous)
                    .stroke(borderColor, lineWidth: 1)
            )
            .kozmosElevation(
                pressed
                    ? KozmosShadows.semanticsElevationRaised
                    : KozmosShadows.semanticsElevationFloating
            )
        }
        .buttonStyle(.plain)
        .disabled(isDisabled)
        .opacity(isDisabled ? 0.5 : 1)
        .animation(reduceMotion ? nil : .easeInOut(duration: 0.3), value: presentation == .labelled)
        .accessibilityLabel(accessibleLabel)
        .accessibilityAddTraits(pressed ? [.isButton, .isSelected] : .isButton)
    }
}

public extension KozmosMapControlButton where Icon == Image {
    /// Convenience initializer for SF Symbol artwork.
    init(
        label: String,
        systemImage: String,
        stateLabel: String? = nil,
        presentation: KozmosMapControlButtonPresentation = .iconOnly,
        emphasis: KozmosMapControlButtonEmphasis = .tinted,
        labelPlacement: KozmosMapControlButtonLabelPlacement = .inline,
        pressed: Bool = false,
        isDisabled: Bool = false,
        action: @escaping () -> Void
    ) {
        self.init(
            label: label,
            stateLabel: stateLabel,
            presentation: presentation,
            emphasis: emphasis,
            labelPlacement: labelPlacement,
            pressed: pressed,
            isDisabled: isDisabled,
            action: action
        ) {
            Image(systemName: systemImage)
        }
    }
}
